src/main/java/ch/njol/skript/sections/EffSecShoot.java

```diff
@@ -0,0 +1,217 @@
+			"(make|let) %livingentities/locations% shoot %entitydatas% [(at|with) (speed|velocity) %-number%] [%-direction%]"
+		);
+		EventValues.registerEventValue(ShootEvent.class, Entity.class, new Getter<Entity, ShootEvent>() {
+			@Override
+			public @Nullable Entity get(ShootEvent shootEvent) {
+				return shootEvent.getProjectile();
+			}
+		}, EventValues.TIME_NOW);
+		EventValues.registerEventValue(ShootEvent.class, Projectile.class, new Getter<Projectile, ShootEvent>() {
+			@Override
+			public @Nullable Projectile get(ShootEvent shootEvent) {
+				return shootEvent.getProjectile() instanceof Projectile projectile ? projectile : null;
+			}
+		}, EventValues.TIME_NOW);
+	}
+
+	private final static Double DEFAULT_SPEED = 5.;
+	private Expression<EntityData<?>> types;
+	private Expression<?> shooters;
+	private @Nullable Expression<Number> velocity;
+	private @Nullable Expression<Direction> direction;
+	public static Entity lastSpawned = null;
+	private @Nullable Trigger trigger;
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult, @Nullable SectionNode sectionNode, @Nullable List<TriggerItem> triggerItems) {
+		types = (Expression<EntityData<?>>) exprs[matchedPattern];
+		shooters = exprs[1 - matchedPattern];
+		velocity = (Expression<Number>) exprs[2];
+		direction = (Expression<Direction>) exprs[3];
+
+		if (sectionNode != null) {
+			AtomicBoolean delayed = new AtomicBoolean(false);
+			Runnable afterLoading = () -> delayed.set(!getParser().getHasDelayBefore().isFalse());
+			trigger = loadCode(sectionNode, "shoot", afterLoading, ShootEvent.class);
+			if (delayed.get()) {
+				Skript.error("Delays cannot be used within a 'shoot' effect section");
+				return false;
+			}
+		}
+		return true;
+	}
+
+	@Override
+	protected @Nullable TriggerItem walk(Event event) {
+		lastSpawned = null;
+		Number finalVelocity = velocity != null ? velocity.getSingle(event) : DEFAULT_SPEED;
+		Direction finalDirection = direction != null ? direction.getSingle(event) : Direction.IDENTITY;
+		if (finalVelocity == null || finalDirection == null)
+			return null;
+
+		for (Object shooter : shooters.getArray(event)) {
+			for (EntityData<?> entityData : types.getArray(event)) {
+				Entity finalProjectile = null;
+				Vector vector;
+				if (shooter instanceof LivingEntity livingShooter) {
+					vector = finalDirection.getDirection(livingShooter.getLocation()).multiply(finalVelocity.doubleValue());
+					Class<? extends Entity> type = entityData.getType();
+					if (Fireball.class.isAssignableFrom(type)) {
+						if (trigger != null) {
+							//noinspection unchecked,rawtypes
+							livingShooter.getWorld().spawn(
+								livingShooter.getEyeLocation().add(vector.clone().normalize().multiply(0.5)),
+								type,
+								(Consumer) afterSpawn(event, vector, entityData, livingShooter)
```

Comment:
you could pull out all the afterSpawn()s and just make a local var at the instanceof LivingEntity check, meaning less duplicate code and fewer warnings to suppress